lib/Driver/ToolChains.cpp

```diff
@@ -1471,6 +1530,266 @@ toolchains::Darwin::constructInvocation(const LinkJobAction &job,
+    llvm_unreachable("invalid link kind");
+  case LinkKind::Executable:
+    // Default case, nothing extra needed.
+    break;
+  case LinkKind::DynamicLibrary:
+    Arguments.push_back("-shared");
+    break;
+  }
+
+  // Select the linker to use.
+  std::string Linker;
+  if (const Arg *A = context.Args.getLastArg(options::OPT_use_ld)) {
+    Linker = A->getValue();
+  }
+  if (!Linker.empty()) {
+    Arguments.push_back(context.Args.MakeArgString("-fuse-ld=" + Linker));
+  }
+
+  // Configure the toolchain.
+  // By default, use the system clang++ to link.
+  const char * Clang = "clang++";
+  if (const Arg *A = context.Args.getLastArg(options::OPT_tools_directory)) {
+    StringRef toolchainPath(A->getValue());
+
+    // If there is a clang in the toolchain folder, use that instead.
+    if (auto toolchainClang = llvm::sys::findProgramByName("clang++", {toolchainPath})) {
+      Clang = context.Args.MakeArgString(toolchainClang.get());
+    }
+
+    // Look for binutils in the toolchain folder.
+    Arguments.push_back("-B");
+    Arguments.push_back(context.Args.MakeArgString(A->getValue()));
+  }
+
+  if (getTriple().getOS() == llvm::Triple::Linux &&
+      job.getKind() == LinkKind::Executable) {
+    Arguments.push_back("-pie");
+  }
+
+  std::string Target = getTriple().str();
+  if (!Target.empty()) {
+    Arguments.push_back("-target");
+    Arguments.push_back(context.Args.MakeArgString(Target));
+  }
+
+  bool staticExecutable = false;
+  bool staticStdlib = false;
+
+  if (context.Args.hasFlag(options::OPT_static_executable,
+                           options::OPT_no_static_executable,
+                           false)) {
+    staticExecutable = true;
+  } else if (context.Args.hasFlag(options::OPT_static_stdlib,
+                                options::OPT_no_static_stdlib,
+                                false)) {
+    staticStdlib = true;
+  }
+
+  SmallString<128> SharedRuntimeLibPath;
+  getRuntimeLibraryPath(SharedRuntimeLibPath, context.Args, *this);
+
+  SmallString<128> StaticRuntimeLibPath;
+  getRuntimeStaticLibraryPath(StaticRuntimeLibPath, context.Args, *this);
+
+  // Add the runtime library link path, which is platform-specific and found
+  // relative to the compiler.
+  if (!(staticExecutable || staticStdlib)) {
+    // FIXME: We probably shouldn't be adding a libpath here unless we know
+    //        ahead of time the standard library won't be copied.
+    Arguments.push_back("-Xlinker");
+    Arguments.push_back(context.Args.MakeArgString("/LIBPATH:" + SharedRuntimeLibPath + "/"
+                          + getTriple().getArchName()));
+  }
+
+  // If we're cross-compiling, we likely already have some library paths set
+  // up through environment variables.
+  // Pass them through to the linker here for convenience.
+  const char *VCToolsInstallDir = ::getenv("VCToolsInstallDir");
+  const char *UniversalCRTSdkDir = ::getenv("UniversalCRTSdkDir");
+  const char *UCRTVersion = ::getenv("UCRTVersion");
```

Comment:
Ugh, this is rather annoying to replicate here, especially since the variables and the paths are subject to Microsoft's control.  We should just request that the user pass in the appropriate paths.

Reply:
The rationale here is that if the user has cross-compiled Swift they'll already have these environment variables set up; it's a nuisance to have to manually copy across `LIB` and `INCLUDE` from an MSVC shell. This eases a significant burden on the user – is it really that bad to have it in another place in the Swift project, at least while cross-compilation is the most reliable Windows toolchain?

Reply:
Actually, thinking about it more, I think you're right here. I've been doing Swift builds regularly so always have those variables set, but someone using a build will just want to have e.g. `INCLUDE` and `LIB` set in their shell configuration file, and it's easy enough to provide instructions on how to do that.